// email-validator-workspace/page.tsx
import { Metadata } from "next";
import fs from "fs";
import path from "path";
import EmailValidatorClient from "./EmailValidatorClient";

// ============================================================================
// UTILITY FUNCTIONS
// ============================================================================

function loadSEOData() {
  try {
    const dataPath = path.join(process.cwd(), 'public', 'data', 'seo.json')
    const fileContent = fs.readFileSync(dataPath, 'utf-8')
    return JSON.parse(fileContent)
  } catch (error) {
    console.error('Error loading SEO data:', error)
    return {}
  }
}

function loadSchemaData() {
  try {
    const dataPath = path.join(process.cwd(), 'public', 'data', 'schema.json')
    const fileContent = fs.readFileSync(dataPath, 'utf-8')
    return JSON.parse(fileContent)
  } catch (error) {
    console.error('Error loading schema data:', error)
    return {}
  }
}

function getSEOData(pageKey) {
  const allSEOData = loadSEOData()
  const keys = pageKey.split('/')
  let data = allSEOData
  for (const key of keys) {
    if (data && data[key]) {
      data = data[key]
    } else {
      return null
    }
  }
  return data
}

function getSchemaData(pageKey) {
  const allSchemaData = loadSchemaData()
  const keys = pageKey.split('/')
  let data = allSchemaData
  for (const key of keys) {
    if (data && data[key]) {
      data = data[key]
    } else {
      return null
    }
  }
  return data
}

// ============================================================================
// METADATA
// ============================================================================

const PAGE_KEY = 'tools/email-validator'

export async function generateMetadata() {
  const seoData = getSEOData(PAGE_KEY)
  
  const baseMetadata = {
    title: seoData?.title || 'Email Validator | Verify & Validate Email Addresses',
    description: seoData?.description || 'Instantly verify email addresses and reduce bounce rates.',
  }

  if (seoData) {
    baseMetadata.keywords = seoData.keywords
    baseMetadata.alternates = { canonical: seoData.canonical }
    baseMetadata.robots = seoData.robots
    baseMetadata.openGraph = {
      title: seoData.ogTitle || seoData.title,
      description: seoData.ogDescription || seoData.description,
      url: seoData.ogUrl,
      images: seoData.ogImage ? [{ url: seoData.ogImage }] : undefined,
    }
    baseMetadata.twitter = {
      card: seoData.twitterCard || 'summary_large_image',
      title: seoData.twitterTitle,
      description: seoData.twitterDescription,
      images: seoData.twitterImage ? [seoData.twitterImage] : undefined,
    }
  }

  return baseMetadata
}

// ============================================================================
// PAGE COMPONENT
// ============================================================================

export default function EmailValidatorPage() {
  const schemaData = getSchemaData(PAGE_KEY)
  
  const schemaScripts = []
  if (schemaData) {
    if (schemaData.webApplication) schemaScripts.push(schemaData.webApplication)
    if (schemaData.breadcrumb) schemaScripts.push(schemaData.breadcrumb)
    if (schemaData.faq) schemaScripts.push(schemaData.faq)
  }

  return (
    <>
      {schemaScripts.map((schema, index) => (
        <script
          key={index}
          type="application/ld+json"
          dangerouslySetInnerHTML={{ __html: JSON.stringify(schema) }}
        />
      ))}
      <EmailValidatorClient />
    </>
  )
}
