// email-validator-workspace/desktop/src/api/client.js
import axios from "axios";

const apiBaseUrl =
  import.meta.env.VITE_API_URL ||
  import.meta.env.REACT_APP_API_URL ||
  "http://localhost:5000/api";
const defaultTimeout = 30000;
const desktopTransport = typeof window !== "undefined" ? window.desktop?.http?.request : null;

let getToken = () => "";
let onTokenRefresh = async () => null;

export function configureApi({ tokenGetter, refreshHandler }) {
  getToken = tokenGetter;
  onTokenRefresh = refreshHandler;
}

function applyAuth(config) {
  const token = getToken();
  if (token) {
    config.headers = { ...(config.headers || {}), Authorization: `Bearer ${token}` };
  }
  return config;
}

function createError(result) {
  const error = new Error(result.message || "Request failed");
  if (result.status) {
    error.response = {
      status: result.status,
      statusText: result.statusText,
      headers: result.headers,
      data: result.data,
    };
  }
  return error;
}

async function send(config) {
  const nextConfig = applyAuth({
    method: "get",
    timeout: defaultTimeout,
    ...config,
  });

  if (desktopTransport) {
    const result = await desktopTransport({
      baseURL: apiBaseUrl,
      method: nextConfig.method,
      url: nextConfig.url,
      params: nextConfig.params,
      data: nextConfig.data,
      headers: nextConfig.headers,
      timeout: nextConfig.timeout,
    });

    if (!result.ok) {
      throw createError(result);
    }

    return {
      data: result.data,
      status: result.status,
      statusText: result.statusText,
      headers: result.headers,
      config: nextConfig,
    };
  }

  return axios({
    baseURL: apiBaseUrl,
    ...nextConfig,
  });
}

async function request(config) {
  try {
    return await send(config);
  } catch (error) {
    if (error?.response?.status === 401 && !config.__retried) {
      config.__retried = true;
      const nextToken = await onTokenRefresh();
      if (nextToken) {
        config.headers = { ...(config.headers || {}), Authorization: `Bearer ${nextToken}` };
        return send(config);
      }
    }

    throw error;
  }
}

const api = {
  request,
  get(url, config = {}) {
    return request({ ...config, method: "get", url });
  },
  post(url, data, config = {}) {
    return request({ ...config, method: "post", url, data });
  },
};

export default api;
