// email-validator-workspace/desktop/src/pages/LoginPage.jsx
import { useState } from "react";
import toast from "react-hot-toast";
import { useAppStore } from "../store/appStore";

export default function LoginPage() {
  const login = useAppStore((s) => s.login);
  const signup = useAppStore((s) => s.signup);
  const openWebsite = useAppStore((s) => s.openWebsite);
  const [mode, setMode] = useState("login");
  const [form, setForm] = useState({
    firstName: "",
    lastName: "",
    email: "",
    password: "",
    confirmPassword: "",
    rememberMe: true,
  });
  const [loading, setLoading] = useState(false);

  const onSubmit = async (e) => {
    e.preventDefault();
    if (mode === "signup" && form.password !== form.confirmPassword) {
      toast.error("Passwords do not match");
      return;
    }
    setLoading(true);
    try {
      if (mode === "signup") {
        await signup(form);
        toast.success("Account created");
      } else {
        await login(form);
        toast.success("Login successful");
      }
    } catch (err) {
      const message = err?.response?.data?.message || (mode === "signup" ? "Signup failed" : "Login failed");
      toast.error(message);
    } finally {
      setLoading(false);
    }
  };

  return (
    <>
      <div className="flex min-h-screen items-center justify-center bg-gradient-to-br from-sky-100 to-emerald-100 px-4 py-6">
        <div className="card w-full max-w-md">
          <div className="mb-4 flex gap-2">
            <button className={`btn ${mode === "login" ? "btn-primary" : "btn-outline"} flex-1`} onClick={() => setMode("login")}>Login</button>
            <button className={`btn ${mode === "signup" ? "btn-primary" : "btn-outline"} flex-1`} onClick={() => setMode("signup")}>Sign Up</button>
          </div>
          <h2 className="mb-2 text-2xl font-bold text-slate-800">{mode === "signup" ? "Create Account" : "Login"}</h2>
          <p className="mb-6 text-sm text-slate-500">
            {mode === "signup" ? "Create your account directly in the app." : "Sign in to access your email validation dashboard."}
          </p>

          <form className="space-y-4" onSubmit={onSubmit}>
            {mode === "signup" && (
              <div className="grid grid-cols-2 gap-3">
                <div>
                  <label className="mb-1 block text-sm font-medium">First Name</label>
                  <input className="input" type="text" value={form.firstName} onChange={(e) => setForm({ ...form, firstName: e.target.value })} />
                </div>
                <div>
                  <label className="mb-1 block text-sm font-medium">Last Name</label>
                  <input className="input" type="text" value={form.lastName} onChange={(e) => setForm({ ...form, lastName: e.target.value })} />
                </div>
              </div>
            )}

            <div>
              <label className="mb-1 block text-sm font-medium">Email</label>
              <input className="input" type="email" value={form.email} onChange={(e) => setForm({ ...form, email: e.target.value })} required />
            </div>

            <div>
              <label className="mb-1 block text-sm font-medium">Password</label>
              <input className="input" type="password" minLength={8} value={form.password} onChange={(e) => setForm({ ...form, password: e.target.value })} required />
            </div>

            {mode === "signup" && (
              <div>
                <label className="mb-1 block text-sm font-medium">Confirm Password</label>
                <input className="input" type="password" minLength={8} value={form.confirmPassword} onChange={(e) => setForm({ ...form, confirmPassword: e.target.value })} required />
              </div>
            )}

            <label className="flex items-center gap-2 text-sm text-slate-600">
              <input type="checkbox" checked={form.rememberMe} onChange={(e) => setForm({ ...form, rememberMe: e.target.checked })} />
              Remember Me
            </label>

            <button type="submit" className="btn-primary w-full" disabled={loading}>
              {loading ? (mode === "signup" ? "Creating account..." : "Logging in...") : (mode === "signup" ? "Create Account" : "Login")}
            </button>
          </form>

          <div className="mt-4 flex items-center justify-between text-sm">
            <button className="text-brand-700 hover:underline" onClick={() => openWebsite("/forgot-password")}>Forgot Password</button>
            <button
              className="text-brand-700 hover:underline"
              onClick={() => setMode(mode === "signup" ? "login" : "signup")}
            >
              {mode === "signup" ? "Already have an account?" : "Create account"}
            </button>
          </div>
        </div>
      </div>
    </>
  );
}
