// email-validator-workspace/desktop/src/store/appStore.js
import { create } from "zustand";
import api, { configureApi } from "../api/client";

const WEBSITE_URL = import.meta.env.WEBSITE_URL || "http://emailnphonelist.com";

export const useAppStore = create((set, get) => ({
  token: "",
  user: null,
  subscription: null,
  online: navigator.onLine,
  settings: {
    rememberMe: false,
    autoUpdate: true,
    saveHistory: true,
  },

  init: async () => {
    const [token, user, rememberMe, autoUpdate, saveHistory] = await Promise.all([
      window.desktop.store.get("authToken"),
      window.desktop.store.get("user"),
      window.desktop.store.get("rememberMe"),
      window.desktop.store.get("autoUpdate"),
      window.desktop.store.get("saveHistory"),
    ]);

    set({
      token: token || "",
      user: user || null,
      settings: {
        rememberMe: !!rememberMe,
        autoUpdate: autoUpdate !== false,
        saveHistory: saveHistory !== false,
      },
    });

    configureApi({
      tokenGetter: () => get().token,
      refreshHandler: async () => {
        const current = get().token;
        if (!current) return "";
        try {
          const response = await api.get("/auth/refresh", {
            headers: { Authorization: `Bearer ${current}` },
          });
          const nextToken = response.data.token;
          set({ token: nextToken, subscription: response.data.subscription || get().subscription });
          await window.desktop.store.set("authToken", nextToken);
          return nextToken;
        } catch (_) {
          set({ token: "", user: null, subscription: null });
          await window.desktop.store.delete("authToken");
          return "";
        }
      },
    });

    window.addEventListener("online", () => set({ online: true }));
    window.addEventListener("offline", () => set({ online: false }));
  },

  login: async ({ email, password, rememberMe }) => {
    const response = await api.post("/auth/login", { email, password });
    const { token, user, subscription } = response.data;
    set({ token, user, subscription, settings: { ...get().settings, rememberMe } });

    await Promise.all([
      window.desktop.store.set("authToken", token),
      window.desktop.store.set("user", user),
      window.desktop.store.set("rememberMe", rememberMe),
    ]);
  },

  signup: async ({ firstName, lastName, email, password, rememberMe }) => {
    const response = await api.post("/auth/signup", { firstName, lastName, email, password });
    const { token, user, subscription } = response.data;
    set({ token, user, subscription, settings: { ...get().settings, rememberMe } });

    await Promise.all([
      window.desktop.store.set("authToken", token),
      window.desktop.store.set("user", user),
      window.desktop.store.set("rememberMe", rememberMe),
    ]);
  },

  logout: async () => {
    try {
      await api.post("/auth/logout");
    } catch (_) {
      // no-op
    }
    set({ token: "", user: null, subscription: null });
    await Promise.all([
      window.desktop.store.delete("authToken"),
      window.desktop.store.delete("user"),
    ]);
  },

  fetchSubscription: async () => {
    const response = await api.get("/user/subscription");
    set({ subscription: response.data });
  },

  updateSettings: async (patch) => {
    const next = { ...get().settings, ...patch };
    set({ settings: next });
    await Promise.all([
      window.desktop.store.set("rememberMe", next.rememberMe),
      window.desktop.store.set("autoUpdate", next.autoUpdate),
      window.desktop.store.set("saveHistory", next.saveHistory),
    ]);
  },

  openWebsite: (path = "") => {
    const url = `${WEBSITE_URL}${path}`;
    window.desktop.openExternal(url);
  },
}));
